sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,563 @@
+either captures of a lambda expression or member variables of a callable
+object.
+
+The primary motivation for this extension is the
+link:../proposed/sycl_ext_oneapi_kernel_compiler.asciidoc[
+sycl_ext_oneapi_kernel_compiler], which allows online compilation of a kernel
+from source code.
+These kernels must have a clean separation between host and device code, and
+they must have a defined order to their arguments because the application sets
+the argument values by their "index" via `handler::set_arg(index, value)`.
+Because variables captured by lambda expressions have no defined order, we need
+some other way to define kernel arguments for use with the
+sycl_ext_oneapi_kernel_compiler extension, and the sycl_ext_oneapi_raw_kernel
+extension provides that mechanism.
+The "raw kernel" feature is a separate extension, though, because it can also
+be used independently from sycl_ext_oneapi_kernel_compiler.
+For example, some users may find it more familiar to define kernels as plain
+functions or they might have other reasons to prefer a clean separation between
+host and device code.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_RAW_KERNEL`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Defining a raw kernel
+
+A raw kernel is a normal C++ function definition, where the function definition
+is decorated with one of the following compile-time properties: `range_kernel`,
+`nd_range_kernel`, or `single_task_kernel`.
+Each of these properties takes a template parameter `Name`, which serves as the
+name of the kernel, similar to the `KernelName` template parameter to
+`parallel_for` for a standard SYCL kernel.
+When a raw kernel has a name, the type used as the name must not be used to
+name any other kernel in the application (either a raw kernel or a "named"
+standard kernel).
+The type used to name a raw kernel has the same limitations as the type used to
+name a standard SYCL kernel as specified in section 5.2 "Naming of kernels" of
+the core SYCL specification.
+A raw kernel may be defined as a function template, however, all instantiations
+of the function must have a unique name, which implies that the type used to
+name the kernel is also somehow templated.
+
+Usually, a raw kernel must have a type-name because the type-name is used to
+obtain a `kernel` object that represents the raw kernel, and a `kernel` object
```

Comment:
Why do need to obtain a `kernel` object when all low-level API use a `kernel` function? 
NOTE: DPC++ compiler today creates a wrapper "kernel function" to create a `kernel` object and lunch it via call operator. I'd like to make sure that we don't create 2 levels of unnecessary abstractions ("wrapper kernel function" + a `kernel` object) to call a raw kernel function.

Reply:
We plan to support this "free function kernel" syntax with the new [sycl_ext_oneapi_kernel_compiler](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc) extension, allowing the application to online compile such kernels from a string.  That extension requires the application to get a `kernel` object in order to enqueue the kernel.  See the example at the bottom of the sycl_ext_oneapi_kernel_compiler spec for details, or see [this example](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler_opencl.asciidoc#example) in the sycl_ext_oneapi_kernel_compiler_opencl extension.

When "free function kernels" are NOT used with the kernel-compiler, we think it would be more convenient in most cases for the user to simply invoke the kernel via the address of the function.  See the conversation starting [here](https://github.com/intel/llvm/pull/11861/files#discussion_r1394593340) for some ideas on how that might work.

Therefore, we probably will support two ways to enqueue these "free function kernels".